質問箱を作成する際にパスワードを設定できるようにする

KishiKyousuke opened the pull request:
ref #5 

## 目的
- 導入予定の質問箱管理画面で行う認証のために、質問箱にパスワードを設定できるようにした
- こちらのPRは質問箱管理画面が完成するまでマージは行いません

## 変更箇所
- 質問箱作成画面にパスワードの入力フォームを追加しました（4文字以上、4文字以下で送信しようとした場合はHTMLから警告が出る）

- パスワードはtext_fieldに入力するようにして、入力したパスワードを確認できるようにしています

- パスワード入力の簡易化のために、確認用パスワードフォームは作成していません

- webpackerを導入し、htmlに直接書かれていたjavascriptのコードをwebpackerでコンパイルされるように`app/javascript`下に移動しました

- 既存のレコードのパスワードカラムが空になってしまうため、デフォルトのパスワードを設定するRakeタスクを作成しました

**質問箱作成画面**
![image](https://user-images.githubusercontent.com/59789739/98318116-3cbd9880-2021-11eb-8d3e-ab3b8ab8e1c6.png)

**パスワード文字数不足**
![image](https://user-images.githubusercontent.com/59789739/98318146-5068ff00-2021-11eb-87cb-82db3ace3214.png)

KishiKyousuke committed:
password_digestカラムをroomsテーブルに追加

gem"bcrypt"を導入しました
質問箱作成時のパスワードを保存するpassword_digestカラムをroomsテーブルに追加しました

KishiKyousuke committed:
質問箱作成・編集フォームにパスワード設定欄を追加

パスワードは4文字以上に制限しました
パスワードが4文字未満または確認用パスワードと一致しない場合はHTMLによってエラーを通知するように実装しました
dbのfixturesにパスワードの項目を追加しました

KishiKyousuke committed:
システムテストにパスワードの入力処理を追加

roomのシステムテストにパスワードの入力の処理を追加しました
testのfixturesにパスワードの項目を追加しました

KishiKyousuke commented:
@machida 
デザインのご確認をよろしくお願いいたします🙇🏻‍♂️

machida commented:
@KishiKyousuke 

ちょっと検討をしたのですが、UIを変更したいと思います。

変更箇所なのなのですが、room作成、編集の

- パスワードを入力中、した後も見れるようにして欲しい（type="text"）で行けるかと思います。
- パスワードの入力は一回だけにして欲しい

この二箇所になります。

ライトなツールであることを伝えたく、パスワード忘れを防ぐことよりも、入力回数の少なさを優先させたく、このようにお願いします🙏

machida commented:
デザインは入れたのでpushしますー
roomsを入れたPRをrebaseしました。なので、一旦このブランチを git pull --rebase origin password-can-be-set-in-question-box をお願いしますー
もしくは、手元のブランチを削除、git fetch origin でも大丈夫です。

machida committed:
room作成フォームの機能追加に伴い全体のデザイン調整を行った

KishiKyousuke committed:
パスワード再入力を撤廃し、入力したパスワードをフォーム上で確認できるように変更

手軽なツールであることを優先するために、パスワードの入力を簡易化しました。
上記の変更に伴い、systemテストを修正いたしました。

komagata commented on app/models/room.rb:
```diff
@@ -3,4 +3,6 @@
 class Room < ApplicationRecord
   has_many :topics, dependent: :destroy
   validates :name, presence: true
+  validates :password, presence: true, length: { minimum: 4 }
```
👍

komagata reviewed (approved):
確認しました、OKですー🙆‍♂️

komagata commented:
完了していたらDraftを外して提出状態にしてください〜

KishiKyousuke commented:
@komagata 
こちらのPRはまだ未修正の段階の #10 のコミットをrebaseしているのですが、一旦こちらをmergeしたあとに #10 の修正後のPRをmergeするという流れでよろしいでしょうか❓

komagata commented:
@KishiKyousuke 

> #10 のコミットをrebaseしているのですが

ってどういう意味でしょうか？

このブランチがadd-room-list-pageの内容をもとにしているとするとadd-room-list-pageがmasterにmergeされた後、今いるブランチをmasterにrebaseしてから完成させる必要があります。

もしくはこのPRの向き先をmasterではなくadd-room-list-page向きにするなどです。

KishiKyousuke commented:
@komagata 
以下の@machidaさんのコメントにあるように、このブランチはデザインを入れていただく段階でadd-room-list-pageをrebaseしているので、add-room-list-pageの内容を元にしていると言えると思います。
現在のadd-room-list-pageの内容はこのブランチにrebaseしてきたときの内容から変更されているため、このブランチにはadd-room-list-pageの古い内容が残されている状態となっています。

![image](https://user-images.githubusercontent.com/59789739/98458875-75ec3900-21d8-11eb-9e18-f2a8fb158fae.png)

komagata commented:
@KishiKyousuke なるほどですね。こちらがマージされたら #10 をマージするで良いと思います。

machida commented:
@KishiKyousuke これを、masterにマージしてしまうと、管理画面がないのにパスワード入力だけがある状態がリリースされてしまうので、このブランチをフィーチャーブランチとして、管理画面系のPRをこのブランチに向けて、管理画面系の機能が揃ったらマスターにマージでいきたいと思いますー

KishiKyousuke commented:
@komagata @machida
となると、先に #10 をマージするということになりますか❓
その場合、 #10 をマージしたmasterをこのブランチにrebaseして、add-room-list-pageの古い内容を更新するというようなことは可能でしょうか❓

machida commented:
@KishiKyousuke 

はい、先に #10 をmasterにマージして、#10 は先にリリースしてしまいますー
#10 をマージしたmasterをこのブランチにrebaseすれば、add-room-list-pageの古い内容も更新されるはずです。

もし、上手く行かなかった場合は、新たにmasterからブランチを作って、このPRに積まれてるcommitを新しいブランチにcherry-pickしましょう。

komagata commented:
@KishiKyousuke すみません、Issueの依存関係がわかっていませんでした。@machida さんのおっしゃる通りで大丈夫です。

KishiKyousuke commented:
既存の質問箱にパスワードを設定するRakeタスクを追加します。

komagata commented on app/views/topics/new.html.slim:
```diff
@@ -32,6 +32,22 @@
       .form__nav
         = link_to '質問収集箱へ', room_path(@room), class: 'form__nav-link a-text-link'
+
+javascript:
```
@KishiKyousuke こちら @machidaさんが追加された箇所だとは思いますが、`app/javascript`以下のwebpackerでコンパイルされて動くように修正お願いしたいです〜

KishiKyousuke committed:
webpackerのインストール

KishiKyousuke committed:
HTML内のJSのコードを`app/javascript`下に移動

rooms及び、topicsのviewのHTML内に書かれていたJSのコードを`app/javascript`下に移動させ、Webpackerでコンパイルされるようにしました。

KishiKyousuke committed:
既存の質問箱にパスワードを設定するRakeタスクを追加

- tallk:oneshot:set_passwordタスクを作成しました

KishiKyousuke commented:
@komagata 
JSのコードの移動、およびRakeタスクの作成を行いました。
`app/javascript`下のディレクトリ構成や、Rakeタスクの名前空間に関しては暫定的なものなのでアドバイスを頂けると幸いです❗
ご確認よろしくお願いいたします🙇🏻‍♂️

komagata commented on app/javascript/packs/application.js:
```diff
@@ -0,0 +1,17 @@
+/* eslint no-console:0 */
```
`app/javascript/packs`以下に置くのはpackされてる（1ファイルにまとめられて使われるJS）ものだけなのでこのぐらいの規模のサイトではapplication.jsだけで大丈夫です。

webpackは利用するJSファイルを最終的に1ファイルにまとめることでキャッシュさせて読み込みを早くするために使われます。

bootcampのソースコードを参考にしてみてください〜

KishiKyousuke commented:
@komagata 
エントリポイントを`application.js`に統一し、JSのコードを`app/javascript`下に移動させました！
ご確認よろしくお願いいたします！

komagata commented on app/javascript/packs/application.js:
```diff
@@ -0,0 +1,18 @@
+// To reference this file, add <%= javascript_pack_tag 'application' %> to the appropriate
+// layout file, like app/views/layouts/application.html.erb
+
+require('../rooms.js')
```
bootcampもこうなっているので変えなきゃなんですが、importの方がいいと思います。

komagata commented on app/javascript/rooms.js:
```diff
@@ -0,0 +1,25 @@
+document.addEventListener('DOMContentLoaded', () => {
```
なんの処理をしているのかという観点の名前にした方がいいと思います。

KishiKyousuke committed:
`views/topics/new.html.slim`のsubmitのclass属性に`is-disabled`を追加

- JSコードの移行によって、フォームに入力がないときでもsubmitボタンが表示されてしまう問題が発生したため

KishiKyousuke commented:
@komagata 
JSファイルを、formの処理・noticeの処理・slideの処理に分割しました！
ご確認よろしくお願いいたします。

KishiKyousuke commented on app/views/rooms/show.html.slim:
```diff
@@ -26,11 +26,16 @@
+      //
+        - unless @room.topics.empty?
+          .form__actions__item
+            = link_to 'トーク開始！', room_topics_path(@room), class: 'a-ex-button is-danger'
```
@machida 
こちらのトーク開始ボタンはコメントアウトしたままでよろしかったでしょうか❓

machida replied:
@KishiKyousuke すいません、このメッセージを見逃してました🙇‍♂️ はい！トーク開始はパスワードでログインした先で表示したいと思いますー

komagata commented on app/javascript/form.js:
```diff
@@ -0,0 +1,17 @@
+document.addEventListener('turbolinks:load', () => {
+  const input = document.getElementById('room_name') || document.getElementById('topic_who');
```
form.jsという名前は全てのformに対して動くという意味合いがあると思うので、何かの個別のformに関する処理は別ファイル（その機能の名前を付けたファイル名）にした方がいいと思います。

KishiKyousuke commented:
@komagata 
これまで`form.js`に集約していたフォームに関する処理を、`room_form.js`と`topic_form.js`に分割しました。
ご確認よろしくお願いいたします。

komagata commented on app/javascript/packs/application.js:
```diff
@@ -0,0 +1,21 @@
+import '../topic_form.js'
+import '../notice.js'
+import '../slide.js'
+
```
不要な空行があるようです〜

komagata commented on app/javascript/room_form.js:
```diff
@@ -0,0 +1,17 @@
+document.addEventListener('turbolinks:load', () => {
```
ファイル名は対象の名前よりも何をやってるのかという観点でつけた方がいいと思います。そしてそれは他のformも共通にできたりしないですかね？

KishiKyousuke commented:
@komagata 
以下の点を修正いたしました。

- 不要な空行の削除
- `topic_form.js`と`room_form.js`を`check_required_input.js`に統一

必須項目としてチェックしたいinput要素に`js-required-input1`と`js-required-input2`というIDを与えることで、その項目に値が入力されている状態のときだけ、submitボタンを押せるようにしています。
ご確認よろしくお願いいたします。

komagata reviewed (approved):
確認しました、OKですー🙆‍♂️
